lib/Sema/CSSolver.cpp

```diff
@@ -922,8 +922,10 @@ void ConstraintSystem::Candidate::applySolutions(
     if (OSR->getDecls().size() == choices.size()) continue;
 
     // Update the expression with the reduced domain.
-    MutableArrayRef<ValueDecl *> decls
-      = TC.Context.AllocateUninitialized<ValueDecl *>(choices.size());
+    MutableArrayRef<ValueDecl *> decls(
+                                Allocator.Allocate<ValueDecl *>(choices.size()),
+                                choices.size());
+
     std::uninitialized_copy(choices.begin(), choices.end(), decls.begin());
     OSR->setDecls(decls);
```

Comment:
If expression is correct there is not going to be any `OSR` expressions in it after `ExprRewriter` because all of the ambiguities have been resolved. But I think the right fix is to allocate to constraint system arena all of the intermediate arrays that `shrink` creates but at the end of the `shrink` phrase re-allocate all of the _final_ arrays to perm arena to make sure that diagnostics are ok.